### PR TITLE
Make performance observers take "type" and "buffered" options without panicking.

### LINKED ISSUE #23225: Intermittent TIMEOUT in /performance-timeline/po-observe.html
```
  ▶ TIMEOUT [expected CRASH] /performance-timeline/po-observe.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ alloc factor 0.010000 0.900000
  └ alloc factor 0.010000 0.900000

  ▶ Unexpected subtest result in /performance-timeline/po-observe.html:
  │ TIMEOUT [expected PASS] PerformanceObserverInit.buffered should retrieve previously buffered entries
  └   → Test timed out
```

### LINKED ISSUE #24781: PerformanceObserverInit doesn't support absence of entryTypes member
This might be an issue with the WebIDL from the spec, but the prose says:
```
entryTypes
    A list of entry types to be observed. If present, the list MUST NOT be empty and all other members MUST NOT be present. Types not recognized by the user agent MUST be ignored. 
```
The /performance-timeline/multiple-buffered-flag-observers.any.html which exercises this reports:
```
  │ [2019-11-19T04:29:53Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js:11:10 Missing required member "entryTypes".
  │ [2019-11-19T04:29:53Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js:11:10 Missing required member "entryTypes".
  │ [2019-11-19T04:29:53Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js:11:10 Missing required member "entryTypes".
  └ [2019-11-19T04:29:53Z ERROR script::dom::bindings::error] Error at http://web-platform.test:8000/performance-timeline/multiple-buffered-flag-observers.any.js:25:10 Missing required member "entryTypes".
```

### LINKED ISSUE #25589: PerformanceObserver option buffered:true causes a double-borrow crash
The spec now only allows `buffered` in combination with the `type` option and not the `entryTypes` option, and since we don't support the `type` option yet, we're failing early in all the WPT tests that use `buffered` and not getting to the actual behavior being tested. This crash thus needs a modified test to detect.

In the course of #24781 I saw this crash and I wanted to make sure I was uncovering a pre-existing bug rather than creating a new one, so I modified performance-timeline/multiple-buffered-flag-observers.any.js. This modified test is non-spec-compliant since it's combining `entryTypes` with `buffered`, but it demonstrates the problem when run on our current master which still allows that combination.

<details><summary>Modified test code</summary>
<pre>
promise_test(() => {
  // The first promise waits for one buffered flag observer to receive 3 entries.
  const promise1 = new Promise(resolve1 => {
    let numObserved1 = 0;
    new PerformanceObserver((entryList, obs) => {
      // This buffered flag observer is constructed after a regular observer detects a mark.
      new PerformanceObserver(list => {
        numObserved1 += list.getEntries().length;
        if (numObserved1 == 3)
          resolve1();
      }).observe({entryTypes: ['mark'], buffered: true});
      obs.disconnect();
    }).observe({entryTypes: ['mark']});
    performance.mark('foo');
  });
  // The second promise waits for another buffered flag observer to receive 3 entries.
  const promise2 = new Promise(resolve2 => {
    step_timeout(() => {
      let numObserved2 = 0;
      // This buffered flag observer is constructed after a delay of 100ms.
      new PerformanceObserver(list => {
        numObserved2 += list.getEntries().length;
        if (numObserved2 == 3)
          resolve2();
      }).observe({entryTypes: ['mark'], buffered: true});
    }, 100);
    performance.mark('bar');
  });
  performance.mark('meow');
  // Pass if and only if both buffered observers received all 3 mark entries.
  return Promise.all([promise1, promise2]);
}, 'Multiple PerformanceObservers with buffered flag see all entries');
</pre>
</details>

It appears that some inappropriate sharing is happening between one PerformanceObserver's entry list and the buffer of entries that another PerformanceObserver is supposed to see, so when one PerformanceObserver tries to clear its list on `.disconnect()` it's re-borrowing an already borrowed value.

### PR DESCRIPTION
**Pull request by pshaughn**
I updated the observe() method to align with spec, and fixed the borrow duration bug @jdm pointed out in #25589 so it wouldn't cause a hard crash. Some tests go from failing to passing, but others go from early failing to later timeout; performance observers still aren't doing the right thing in all cases.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix #24781 and fix #25589 and fix #23225

- [X] There are tests for these changes

### CODE REVIEW

**Comment by pshaughn** [comment]
If the timeouts in CI are the same as in my local environment, they'll need a followup issue.

**Comment by jdm** [comment]
<details>

```
10 unexpected results that are NOT known-intermittents:
  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.html:
  â”” PASS [expected FAIL] Calling observe() with entryTypes and then type should throw an InvalidModificationError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.html:
  â”” PASS [expected FAIL] Calling observe() with type and then entryTypes should throw an InvalidModificationError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.html:
  â”” PASS [expected FAIL] Calling observe() with type and entryTypes should throw a SyntaxError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.html:
  â”” PASS [expected FAIL] Passing in unknown values to type does throw an exception.

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.html:
  â”‚ TIMEOUT [expected FAIL] observe() with different type values stacks.
  â””   â†’ Test timed out


  â–¶ TIMEOUT [expected OK] /performance-timeline/multiple-buffered-flag-observers.any.worker.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”” 

  â–¶ Unexpected subtest result in /performance-timeline/multiple-buffered-flag-observers.any.worker.html:
  â”‚ TIMEOUT [expected PASS] Multiple PerformanceObservers with buffered flag see all entries
  â””   â†’ Test timed out


  â–¶ Unexpected subtest result in /performance-timeline/buffered-flag-observer.any.worker.html:
  â”‚ TIMEOUT [expected FAIL] PerformanceObserver with buffered flag should see past and future entries.
  â””   â†’ Test timed out


  â–¶ Unexpected subtest result in /performance-timeline/buffered-flag-observer.any.html:
  â”‚ TIMEOUT [expected FAIL] PerformanceObserver with buffered flag should see past and future entries.
  â””   â†’ Test timed out


  â–¶ TIMEOUT [expected OK] /performance-timeline/multiple-buffered-flag-observers.any.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”” 

  â–¶ Unexpected subtest result in /performance-timeline/multiple-buffered-flag-observers.any.html:
  â”‚ TIMEOUT [expected PASS] Multiple PerformanceObservers with buffered flag see all entries
  â””   â†’ Test timed out


  â–¶ Unexpected subtest result in /performance-timeline/buffered-flag-after-timeout.any.worker.html:
  â”‚ TIMEOUT [expected FAIL] PerformanceObserver with buffered flag sees entry after timeout
  â””   â†’ Test timed out


  â–¶ TIMEOUT [expected OK] /performance-timeline/buffered-flag-after-timeout.any.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”” 

  â–¶ Unexpected subtest result in /performance-timeline/buffered-flag-after-timeout.any.html:
  â”‚ TIMEOUT [expected PASS] PerformanceObserver with buffered flag sees entry after timeout
  â””   â†’ Test timed out


  â–¶ TIMEOUT [expected ERROR] /resource-timing/buffered-flag.any.worker.html
  â”‚ 
  â”‚ 
  â”‚ 
  â”” 


  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.worker.html:
  â”” PASS [expected FAIL] Calling observe() with entryTypes and then type should throw an InvalidModificationError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.worker.html:
  â”” PASS [expected FAIL] Calling observe() with type and then entryTypes should throw an InvalidModificationError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.worker.html:
  â”” PASS [expected FAIL] Calling observe() with type and entryTypes should throw a SyntaxError

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.worker.html:
  â”” PASS [expected FAIL] Passing in unknown values to type does throw an exception.

  â–¶ Unexpected subtest result in /performance-timeline/po-observe-type.any.worker.html:
  â”‚ TIMEOUT [expected FAIL] observe() with different type values stacks.
  â””   â†’ Test timed out


  â–¶ OK [expected ERROR] /resource-timing/buffered-flag.any.html

  â–¶ Unexpected subtest result in /resource-timing/buffered-flag.any.html:
  â”‚ FAIL [expected TIMEOUT] PerformanceObserver with buffered flag sees previous resource entries.
  â”‚   â†’ assert_equals: There should be 1 resource entry. expected 1 but got 2
  â”‚ 
  â”‚ @http://web-platform.test:8000/resource-timing/buffered-flag.any.js:9:7
  â”‚ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  â”” Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2065:32

```

</details>

**Comment by pshaughn** [comment]
I decided to do a slightly off-spec thing the tests were implying in order to avoid intermittency; I'm expecting this to have a consistent result now.
@bors-servo try=wpt

**Comment by jdm** [comment]
```
  ▶ OK [expected ERROR] /resource-timing/buffered-flag.any.worker.html
  ▶ Unexpected subtest result in /resource-timing/buffered-flag.any.worker.html:
  │ FAIL [expected TIMEOUT] PerformanceObserver with buffered flag sees previous resource entries.
  │   → assert_true: expected true got false
  │ 
  │ @http://web-platform.test:8000/resource-timing/buffered-flag.any.js:14:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2065:32
  ▶ OK [expected ERROR] /resource-timing/buffered-flag.any.html
  ▶ Unexpected subtest result in /resource-timing/buffered-flag.any.html:
  │ FAIL [expected TIMEOUT] PerformanceObserver with buffered flag sees previous resource entries.
  │   → assert_equals: There should be 1 resource entry. expected 1 but got 2
  │ 
  │ @http://web-platform.test:8000/resource-timing/buffered-flag.any.js:9:7
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:2024:25
  └ Test.prototype.step_func_done/<@http://web-platform.test:8000/resources/testharness.js:2065:32
```

**Comment by pshaughn** [comment]
Interesting! Now that we'rd not getting a timeout there, we see that the script is buffering a resource entry for its _own_ url!

File: components/script/dom/performanceobserver.rs [hunk 1]
```diff
@@ -106,31 +115,114 @@ impl PerformanceObserver {
...
-    // https://w3c.github.io/performance-timeline/#dom-performanceobserver-disconnect()
+    // https://w3c.github.io/performance-timeline/#dom-performanceobserver-disconnect
     fn Disconnect(&self) {
         self.global().performance().remove_observer(self);
         self.entries.borrow_mut().clear();
     }
+
+    // https://w3c.github.io/performance-timeline/#takerecords-method
+    fn TakeRecords(&self) -> Vec<DomRoot<PerformanceEntry>> {
+        warn!("About to borrow_mut entries to copy and clear them in TakeRecords");
```
Comment (pshaughn) [inline components/script/dom/performanceobserver.rs]: This was left over from bug-hunting, I will remove it

**Comment by pshaughn** [comment]
#21067

**Comment by pshaughn** [comment]
good old #24726, I'm gonna miss that bug when it's gone

**Comment by jdm** [comment]
@bors-servo r+
Thanks for fixing this!

**Pull request merged**